File: pset6/credit.py
from cs50 import get_int

def main():
    ccnum = get_int("Enter Credit card number: ")
    digits = len(str(ccnum))

    if digits == 16 or digits == 13 or digits == 15:
        con = check_validity(ccnum, digits)
    else:
        con = False

    if con:
        if str(ccnum)[0] in ['4']:
            print("VISA")
        elif str(ccnum)[:2] in ['51', '52', '53', '54', '55']:
            print("MASTERCARD")
        elif str(ccnum)[:2] in ['34', '37']:
            print('AMEX')
        else:
            print("INVALID")
    else:
        print("INVALID")


def check_validity(num, dig):
    sum1=0
    sum2=0
    flag =0
    flag2=0
    count=1
    for i in range(dig - 1, -1, -1):
        flag = int(str(num)[i])
        if (count) % 2 == 0:
            flag2 = flag * 2
            if flag2 > 9:
                tmp = int(str(flag2)[0]) + int(str(flag2)[1])
                flag2 = tmp
            sum1 = sum1 + flag2
        else:
            sum2 = sum2 + flag
        count+=1
    final_num = sum1 + sum2
    if final_num % 10 ==0:
        return True
    else:
        return False



if __name__=='__main__':
    main()

File: pset1/credit.c
#include <stdio.h>
#include <math.h>
#include <cs50.h>

bool check_validity(long, int);
int count_num_of_digits(long);

int main(void)
{
    long ccnum = get_long("Enter credit card number: ");
    int digits = count_num_of_digits(ccnum);
    bool con;
    if (digits == 16 || digits == 13 || digits == 15)
    {
       con = check_validity(ccnum, digits);
    }
    else
    {
        con = false;
    }

    if (con)
    {
        long res = pow(10, digits-2);
        int num_to_check = ccnum / res;
        if (num_to_check / 10 == 4)
        {
            printf("VISA\n");
        }
        else if (num_to_check == 51 || num_to_check == 52 || num_to_check == 53 || num_to_check == 54 || num_to_check == 55)
        {
            printf("MASTERCARD\n");
        }
        else if (num_to_check == 34 || num_to_check == 37)
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}


int count_num_of_digits(long num)
{
    int count = 0;
    while (num != 0)
    {
        num = num /10;
        count++;
    }
    return count;
}


bool check_validity(long num, int dig)
{
    int sum1 = 0;
    int sum2 = 0;
    int flag = 0;
    int flag2 = 0;
    for (int i = 1; i <= dig; i++ )
    {
        flag = num % 10;
        if (i % 2 == 0)
        {
            flag2 = flag * 2;
            if (flag2 > 9)
            {
                int dum = flag2 % 10;
                int new_num = flag2 / 10;
                flag2 = dum + new_num;
            }
            sum1 = sum1 + flag2;
        }
        else
        {
            sum2 = sum2 + flag;
        }
        num = num / 10;
    }
    int final_num = sum1 + sum2;

    if (final_num % 10 == 0)
    {
        return true;
    }
    else
    {
        return false;
    }

}


